crates/node/src/lib.rs

```diff
@@ -986,6 +1037,118 @@ impl Node {
 
         Ok(())
     }
+
+    async fn perform_catchup(
+        &mut self,
+        context_id: calimero_primitives::context::ContextId,
+        chosen_peer: libp2p::PeerId,
+    ) -> eyre::Result<()> {
+        let handle = self.store.handle();
+
+        let (mut context, request) =
+            match handle.get(&calimero_store::key::ContextMeta::new(context_id))? {
+                Some(ctx_meta) => (
+                    Some(calimero_primitives::context::Context {
+                        id: context_id,
+                        application_id: ctx_meta.application_id.clone().into_string().into(),
+                    }),
+                    types::CatchupRequest {
+                        context_id,
+                        application_id: Some(ctx_meta.application_id.clone().into_string().into()),
+                        last_executed_transaction_hash: ctx_meta.last_transaction_hash.into(),
+                        batch_size: self.network_client.catchup_config.batch_size,
+                    },
+                ),
+                None => (
+                    None,
+                    types::CatchupRequest {
+                        context_id,
+                        application_id: None,
+                        last_executed_transaction_hash: calimero_primitives::hash::Hash::default(),
+                        batch_size: self.network_client.catchup_config.batch_size,
+                    },
+                ),
+            };
+
+        let mut stream = self.network_client.open_stream(chosen_peer).await?;
+
+        let request = serde_json::to_vec(&types::CatchupStreamMessage::Request(request))?;
+
+        stream
+            .send(calimero_network::stream::Message { data: request })
+            .await?;
+
+        while let Some(message) = stream.next().await {
+            match serde_json::from_slice(&message?.data)? {
+                types::CatchupStreamMessage::TransactionsBatch(response) => {
```

Comment:
btw, this could be `CatchupError`, it's not handled

Reply:
in that case, these:

https://github.com/calimero-network/core/blob/1631e492df629569b86086756efbcef4953a04ba/crates/node/src/lib.rs#L926-L931

https://github.com/calimero-network/core/blob/1631e492df629569b86086756efbcef4953a04ba/crates/node/src/lib.rs#L939-L944

https://github.com/calimero-network/core/blob/1631e492df629569b86086756efbcef4953a04ba/crates/node/src/catchup.rs#L63-L70

have wrong implementations, right?

Reply:
that's why it's good to rely on the type system to guide these things, I'll patch stream to be generic over a `codec` like `MessageJsonCodec` which would then be generic over an serializable type